### PR TITLE
add approximate zCDP

### LINKED ISSUE #2005: Add δ-approximate zCDP

### PR DESCRIPTION
**Pull request by Shoeboxam**
Close #2005

### CODE REVIEW

**Comment by Shoeboxam** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/opendp/opendp/2008?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#2008** <a href="https://app.graphite.dev/github/pr/opendp/opendp/2008?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#1768** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1768?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 2 other dependent PRs ([#1854](https://github.com/opendp/opendp/pull/1854) <a href="https://app.graphite.dev/github/pr/opendp/opendp/1854?utm_source=stack-comment-icon" target="_blank">Graphite</a>, [#1999](https://github.com/opendp/opendp/pull/1999) <a href="https://app.graphite.dev/github/pr/opendp/opendp/1999?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#2007** <a href="https://app.graphite.dev/github/pr/opendp/opendp/2007?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1998** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1998?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @Shoeboxam and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: python/test/context/test_context.py [hunk 1]
```diff
@@ -237,6 +241,27 @@ def test_rho_to_eps():
...
+    context_azcdp = context.query().compositor(3, output_measure=azcdp_measure, alpha=0.3).release()
+
+    dom, met = context_azcdp.accountant.input_space
+    context_azcdp(dp.m.make_user_measurement(
+        dom, met,
+        azcdp_measure,
+        lambda x: x,
+        # remaining rho, and catastrophic delta
+        lambda _: (.006, 1e-6 * .3 / 3)
+    ))
```
Comment (mccalluc) [inline python/test/context/test_context.py:262]: Make some kind of assertion? Maybe just about the shape of the response, if the actual values are indeterminate.

File: rust/src/combinators/measure_cast/zCDP_to_approxDP/test.rs [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+        AbsoluteDistance::<f64>::default(),
+        1.,
+        None,
+    )?;
+
+    let m_adp = make_zCDP_to_approxDP(m_zcdp)?;
+
+    let curve = m_adp.map(&1.0)?;
+    let epsilon = curve.epsilon(&1e-7)?;
+    assert_eq!(epsilon, 5.6708588354687715);
```
Comment (mccalluc) [inline rust/src/combinators/measure_cast/zCDP_to_approxDP/test.rs:21]: Explain value?

**Review by mccalluc — COMMENTED** [review]
Maybe a stronger assertion in `test_approx_to_zCDP`? Apart from that, I think it's mostly just comments. Re-review and approval should be quick, but let me know if there's anything particular to look at.

File: python/src/opendp/context.py [hunk 3]
```diff
@@ -260,19 +261,22 @@ def range_warning(name, value, info_level, warn_level):
...
+        range_warning('epsilon', epsilon, 1, 5)
+        measure, loss = max_divergence(), epsilon
+
+    if rho is not None:
+        range_warning('rho', rho, 0.25, 0.5)
+        measure, loss = zero_concentrated_divergence(), rho
+    
+    if delta is not None:
+        range_warning('delta', delta, 1e-6, 1e-6)
+        measure, loss = approximate(measure), (loss, delta) # type: ignore[assignment]
```
[the remark this answers is not in the document]
Comment (Shoeboxam) [reply python/src/opendp/context.py:277]: I'll just return early here.

File: python/src/opendp/extras/polars/__init__.py [hunk 4]
```diff
@@ -536,7 +535,7 @@ def make(scale, threshold=None):
                 )
             
             # when the output measure is δ-approximate, then there are two free parameters to tune
-            if query._output_measure == fixed_smoothed_max_divergence():
+            if str(query._output_measure.type).startswith("Approximate"):
```
[the remark this answers is not in the document]
Comment (Shoeboxam) [reply python/src/opendp/extras/polars/__init__.py:538]: I used a getattr trick. Sometime when we replace `RuntimeType | str` with `RuntimeType` then we'll just access origin.

**Review by mccalluc — DISMISSED** [review]
Just doing a quick scan, but this seem ok. (It's only showing as one commit, so it's harder for me to spot the changes.)

**Pull request merged**